api/token.go

```diff
@@ -600,22 +584,63 @@ func (a *API) issueRefreshToken(ctx context.Context, conn *storage.Connection, u
 		if terr != nil {
 			return terr
 		}
-		terr = models.AddClaimToSession(tx, session, signInMethod)
+		terr = models.AddClaimToSession(tx, session, authenticationMethod)
 		if terr != nil {
 			return terr
 		}
 
-		if isMFASignInMethod {
-			if err := session.UpdateAssociatedFactor(tx, grantParams.FactorID); err != nil {
-				return err
-			}
+		tokenString, terr = generateAccessToken(tx, user, refreshToken.SessionId, time.Second*time.Duration(config.JWT.Exp), config.JWT.Secret)
+		if terr != nil {
+			return internalServerError("error generating jwt token").WithInternalError(terr)
 		}
-		_, aal := calculateAALFromClaims(currentClaims, signInMethod)
-		if err := session.UpdateAAL(tx, aal); err != nil {
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &AccessTokenResponse{
+		Token:        tokenString,
+		TokenType:    "bearer",
+		ExpiresIn:    config.JWT.Exp,
+		RefreshToken: refreshToken.Token,
+		User:         user,
+	}, nil
+}
+
+func (a *API) updateMFASessionAndClaims(ctx context.Context, conn *storage.Connection, user *models.User, authenticationMethod models.AuthenticationMethod, grantParams models.GrantParams) (*AccessTokenResponse, error) {
+	config := a.config
+	var tokenString string
+	var refreshToken *models.RefreshToken
+	currentClaims := getClaims(ctx)
+	err := conn.Transaction(func(tx *storage.Connection) error {
+		sessionId, terr := uuid.FromString(currentClaims.SessionId)
+		if terr != nil {
+			return internalServerError("Cannot read SessionId claim as UUID").WithInternalError(terr)
+		}
```

Comment:
Can you move this validation logic outside of the transaction. It's a good idea to keep transactions as short as possible as they use up a key resource -- database connections.